icepyx/core/types.py

```diff
@@ -0,0 +1,79 @@
+from __future__ import annotations
+
+from typing import Literal, TypedDict
+from typing_extensions import NotRequired
+
+
+CMRParamsBase = TypedDict(
+    "CMRParamsBase",
+    {
+        "temporal": NotRequired[str],
+        "options[readable_granule_name][pattern]": NotRequired[str],
+        "options[spatial][or]": NotRequired[str],
+        "readable_granule_name[]": NotRequired[str],
+    },
+)
```

Comment:
I'm not sure I fully understand why:

```
class CMRParamsBase(TypedDict):
        temporal: NotRequired[str],
        options[readable_granule_name][pattern]: NotRequired[str],
        options[spatial][or]: NotRequired[str],
        readable_granule_name[]: NotRequired[str],
```

(the "new" way) isn't an option?

> I'm shooting for annotating the code as-is to enable deeper future refactors with less mistakes. I would prefer to pass around dataclasses :)

Totally makes sense. Still learning about dataclasses so can't agree with you yet!

Reply:
Good question! Python will try to access objects named `options` (to call `__getitem__` IIRC), then `readable_granule_name`, and `pattern` to evaluate that code and you'll get an error:

```python
>>> class Foo(TypedDict):
...   quz[qux]: int
... 
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "<stdin>", line 2, in Foo
NameError: name 'quz' is not defined
```

Class attributes must follow naming rules, e.g. can't start with a number or symbol, can't use certain symbols, etc. I'm struggling to find the doc in the time I have right now :X

Reply:
Got it - thanks! [Docs](https://docs.python.org/3.12/library/typing.html#typing.TypedDict) (turns out I had them open).